### DISCUSSION TITLE
Is it possible to have multiple draft versions?

### DISCUSSION THREAD

**Discussion by mariojsnunes** [OrchardCMS/OrchardCore#17096] [Q&A] [answered]
I'm trying to create a version programmatically, every time an editor saves an item.
A new version is created only if the current latest is published, not if it's draft.
But I'd like to create new draft versions even if the latest is a draft.

I've tried with both
`await contentManager.CreateAsync(contentItem, VersionOptions.DraftRequired);` and
`await contentManager.CreateAsync(contentItem, VersionOptions.Draft);`
But the behavior is the same.

Edit: I'm on version 1.8.3

**Comment by Piedone** [comment, marked as answer]
No, not exactly like this, since this is not how OC's versioning is designed. However, there are a couple of workarounds you can look into:

- From a handler, when a draft is saved, publish it, then on the next line create a new draft and set the former published version as published (you can do this by setting the `Published` flag on the item directly, bypassing the content manager).
- If you don't need multiple draft versions, per se, and just want to see how the drafts changed, Audit Trail already records these. E:g. these below were all just draft changes, and under Details you can see the diff with the previous draft:

![{5CBF47EA-A025-4E34-84B7-6B3CAEABDC36}](https://github.com/user-attachments/assets/157dbf0a-69a3-480a-a033-6bc163c0cd41)

What's your use case, though?

**Reply by mariojsnunes** [reply]
It's an intranet where users can create articles. We provide our own UI, and users can't access the Orchard backoffice.
We already have a feature where users can see previous versions. But just this week a client asked us to rollback a draft version that had disappeared and this was the cause. (We used the audit trail to recover it as you mentioned)

The first option looks better for us, although it's a bit hacky :)
Thanks!

**Reply by mariojsnunes** [reply]
Been trying to make this work, still without success.

These were my assumptions:
1. "From a handler, when a draft is saved" means to use contentManager.SaveDraftAsync() and have a ContentPartHandler which overrides the DraftSavedAsync event.
2. Publish the draft - by setting Published to true and saving with session.SaveAsync.
3. Create a new Draft - tried with both contentManager and session.SaveAsync of the same contentItem
  i. Also tried setting the item id to 0 and generating a new versionId, before session.SaveAsync
5. Republish the original published version - by setting Published to true and saving with session.SaveAsync

For step 2, wouldn't want to use contentManager.PublishAsync as it would trigger other events.

Is this what you had in mind? Thanks!

**Reply by Piedone** [reply]
1. Yes. You can also check `UpdatedAsync()` in the handler, though that'll run on every editor update, including ones that resulted in a validation error, and that were part of a publish.
2. For the first (but not the second) publish, I meant with `IContentManager.PublishAsync()` but if it works then yes, just setting the flag is better to bypass and handler that we'd mislead.
3. You can create a draft with `IContentManager.GetAsync(ID, VersionOptions.IsDraftRequired)`.
4. Yep.
